### PR TITLE
getJobs 100% coverage

### PR DESCRIPTION
**Pull request by randallm**

### CODE REVIEW

File: lib/queue.js [hunk 1]
```diff
@@ -260,51 +257,52 @@ Queue.prototype._scanForJobs = function (cursor, set, cb) {
...
-Queue.prototype._getJobs = function (queueType, start, end, cb) {
+Queue.prototype.getJobs = function (type, page, cb) {
   const promise = this._commandable().then((client) => {
     const idsPromise = helpers.deferred();
-    switch (queueType) {
+    switch (type) {
       case 'failed':
       case 'succeeded':
-        this._scanForJobs('0', new Set(), idsPromise.defer());
+        client.scard(this.toKey(type), () => {
```
Comment (skeggse) [inline lib/queue.js]: :)

File: lib/queue.js [hunk 2]
```diff
@@ -260,51 +257,52 @@ Queue.prototype._scanForJobs = function (cursor, set, cb) {
...
     }
 
-    this._scanForJobs(nextCursor, set, cb);
+    this._scanForJobs(key, nextCursor, size, set, cb);
   });
 };
 
-// TODO(randall): Implement (start, end) parameters
-Queue.prototype._getJobs = function (queueType, start, end, cb) {
+Queue.prototype.getJobs = function (type, page, cb) {
```
Comment (skeggse) [inline lib/queue.js]: Can you add a doc comment that describes the `page` object?

File: test/queue-test.js [hunk 3]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
     });
   });
 
+  it.describe('getJobs', (it) => {
+    it('gets waiting jobs', async (t) => {
+      const queue = t.context.makeQueue();
+
+      const insertedJobs = [];
+      for (let i = 0; i < 3; i++) {
+        let job = await queue.createJob({foo: 'bar'}).save();
```
Comment (skeggse) [inline test/queue-test.js]: `const job`

File: test/queue-test.js [hunk 4]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+        insertedJobs.push(job.toData());
+      }
+
+      const jobs = await queue.getJobs('waiting', {start: 0, end: 2});
+      for (let job of jobs) {
+        t.true(insertedJobs.includes(job.toData()));
+      }
```
Comment (skeggse) [inline test/queue-test.js]: Make `insertedJobs` and `jobs` a `Set`, and you can just use `t.deepEqual` :)

File: test/queue-test.js [hunk 5]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+    });
+
+    it('gets active jobs', async (t) => {
+      const queue = t.context.makeQueue();
+
+      await queue.createJob({foo: 'bar'}).save();
+
+      queue.process(async (job) => {
+        const jobs = await queue.getJobs('active', {start: 0, end: 100});
+        t.is(jobs[0].toData(), job.toData());
```
Comment (skeggse) [inline test/queue-test.js]: I'd prefer that this check the `id` instead of the data

File: test/queue-test.js (hunk 4 above)
Comment (skeggse) [reply test/queue-test.js]: I'd prefer that this check the `id` instead of the data, if possible - it seems that this won't be a good test because they will all have the same data

File: test/queue-test.js [hunk 6]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+      });
+
+      await helpers.waitOn(queue, 'succeeded');
```
Comment (skeggse) [inline test/queue-test.js]: `true` for error handling

File: test/queue-test.js [hunk 7]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+    });
+
+    it('gets delayed jobs', async (t) => {
+      const queue = t.context.makeQueue();
+
+      queue.process(async () => {});
+
+      const job = await queue.createJob({foo: 'bar'})
+        .delayUntil(Date.now() + 1000)
```
Comment (skeggse) [inline test/queue-test.js]: Increase this to like `100000` - doesn't hurt, and these tests are pretty intensive

File: test/queue-test.js [hunk 8]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+      const queue = t.context.makeQueue();
+
+      queue.process(async () => {});
+
+      const job = await queue.createJob({foo: 'bar'}).save();
+
+      await helpers.waitOn(queue, 'succeeded');
+
+      const jobs = await queue.getJobs('succeeded', {size: 1});
+      t.not(jobs[0].toData(), job.toData());
```
Comment (skeggse) [inline test/queue-test.js]: Why is this the case?

File: test/queue-test.js [hunk 9]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+      });
+
+      queue.process(async () => {});
+
+      // Choose a big number for numbers of jobs created, because otherwise the
+      // set will be encoded as an intset and SSCAN will ignore the COUNT param.
+      // https://redis.io/commands/scan#the-count-option
+      for (let i = 0; i < 10000; i++) {
+        await queue.createJob({foo: 'bar'}).save();
+      }
```
Comment (skeggse) [inline test/queue-test.js]: How long does this test take? Hopefully it doesn't add to the time the tests take to run. You might consider doing this with `Promise.all` so the redis commands all flush concurrently

File: test/queue-test.js [hunk 10]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+
+      const jobs = await queue.getJobs('succeeded', {size: 80});
+      t.is(jobs.length, 80);
```
Comment (skeggse) [inline test/queue-test.js]: Maybe also test that there aren't duplicates? I know this won't happen in your implementation, but the implementation may change at some future date :)

File: test/queue-test.js [hunk 11]
```diff
@@ -572,6 +572,126 @@ describe('Queue', (it) => {
...
+    it('accepts size parameter for set types', async (t) => {
+      const queue = t.context.makeQueue();
+
+      await t.notThrows(queue.getJobs('succeeded', {size: 10}));
+    });
+
+    it('rejects improper queue type', async (t) => {
+      const queue = t.context.makeQueue();
+
+      await t.throws(queue.getJobs('not-a-queue-type'), 'Improper queue type');
```
Comment (skeggse) [inline test/queue-test.js]: maybe `/improper queue type/i`

File: lib/queue.js [hunk 12]
```diff
@@ -243,12 +243,9 @@ Queue.prototype.checkHealth = function (cb) {
...
 };
 
-Queue.prototype._scanForJobs = function (cursor, set, cb) {
-  // Ideally, we would be able to paginate redis sets like we do sorted ranges.
-  // For an MVP, however, we're aiming to replace our usage of Arena for Bull,
-  // and furthermore Bull 1.x calls SMEMBERS -- which is more blocking than SSCAN.
-
-  this.client.sscan(this.toKey('failed'), cursor, 'COUNT', this.settings.redisScanCount, (err, results) => {
+Queue.prototype._scanForJobs = function (key, cursor, size, set, cb) {
+  const count = Math.min(size, this.settings.redisScanCount);
```
Comment (skeggse) [inline lib/queue.js]: Maybe give this a more descriptive name like `batchCount`

File: lib/queue.js [hunk 13]
```diff
@@ -243,12 +243,9 @@ Queue.prototype.checkHealth = function (cb) {
   return promise;
 };
 
-Queue.prototype._scanForJobs = function (cursor, set, cb) {
-  // Ideally, we would be able to paginate redis sets like we do sorted ranges.
-  // For an MVP, however, we're aiming to replace our usage of Arena for Bull,
-  // and furthermore Bull 1.x calls SMEMBERS -- which is more blocking than SSCAN.
-
-  this.client.sscan(this.toKey('failed'), cursor, 'COUNT', this.settings.redisScanCount, (err, results) => {
+Queue.prototype._scanForJobs = function (key, cursor, size, set, cb) {
```
Comment (skeggse) [inline lib/queue.js:246]: It's a little strange that this accepts a set as an input, but produces an array. I'd be tempted to have the caller convert to an array instead.

**Review by skeggse — CHANGES_REQUESTED** [review]

**Comment by skeggse** [comment]
:fire:

File: test/queue-test.js (hunk 8 above)
Comment (randallm) [reply test/queue-test.js]: Removing this line because it tests behavior unrelated to `getJobs`, but for reference it's because the status changes from "created" to "succeeded".

File: test/queue-test.js (hunk 4 above)
Comment (randallm) [reply test/queue-test.js]: fyi on the first review point, Ava's [`deepEqual` implementation](https://github.com/avajs/ava#deepequalvalue-expected-message) compares sets in-order so deep-equal will not actually work as intended in this scenario.

**Comment by randallm** [comment]
Updated!

File: lib/queue.js [hunk 14]
```diff
@@ -265,23 +265,32 @@ Queue.prototype._scanForJobs = function (key, cursor, size, set, cb) {
...
+/**
+ * Get jobs from queue type.
+ *
+ * @param {String} type The queue type ("failed", "succeeded", "waiting", etc.)
+ * @param {Object} page
+ * @param {Number=} page.start Start of query range for waiting/active/delayed queue types
+ * @param {Number=} page.end End of query range for waiting/active/delayed queue types
+ * @param {Number=} page.size Number jobs to return for failed/succeeded (SET) types
+ * @param {Function=} callback Called with the equivalent of the returned promise.
+ * @return {Promise<Array>} Resolves to the jobs the function found.
```
Comment (skeggse) [inline lib/queue.js]: Put `@returns {Promise<Job[]>}`

File: test/queue-test.js [hunk 15]
```diff
@@ -648,17 +640,31 @@ describe('Queue', (it) => {
...
 
       // Choose a big number for numbers of jobs created, because otherwise the
       // set will be encoded as an intset and SSCAN will ignore the COUNT param.
       // https://redis.io/commands/scan#the-count-option
+      const createJobPromises = [];
       for (let i = 0; i < 10000; i++) {
-        await queue.createJob({foo: 'bar'}).save();
+        createJobPromises.push(queue.createJob({foo: 'bar'}).save());
       }
+      Promise.all(createJobPromises);
```
Comment (skeggse) [inline test/queue-test.js]: you should `await` this, otherwise it'll process asynchronously (and you should probably define the `process` before creating the jobs)

File: test/queue-test.js [hunk 16]
```diff
@@ -648,17 +640,31 @@ describe('Queue', (it) => {
...
+
+      // Wait for all jobs to process to make sure the SET encoding is a hash table
+      // rather than an intset.
+      const allJobsProcessed = new Promise((resolve) => {
```
Comment (skeggse) [inline test/queue-test.js]: switch this to use `reef` (there are examples of it being used with `process`) - ask me if you need clarification

**Review by skeggse — CHANGES_REQUESTED** [review]
🙏  last few then lgtm

**Comment by randallm** [comment]
Updated @skeggse

**Review by skeggse — APPROVED** [review]

File: test/queue-test.js (hunk 4 above)
Comment (skeggse) [reply test/queue-test.js]: Wait...why, though :O
